sda-commons-server-s3-testing/src/main/java/org/sdase/commons/server/s3/testing/S3ClassExtension.java

```diff
@@ -1,63 +1,125 @@
+import software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3Client;
 
 /**
- * JUnit 5 extension for running a AWS S3-compatible object storage alongside the (integration)
+ * JUnit 5 extension for running an AWS S3-compatible object storage alongside the (integration)
  * tests. Use {@link #getEndpoint()} to retrieve the endpoint URL to connect to.
  *
  * <p>Example usage:
  *
  * <pre>
+ * &#64;LocalS3
+ * class MyS3Test {
+ *
  *   &#64;RegisterExtension
- *   static final S3ClassExtension S3_EXTENSION = S3ClassExtension.builder().build();
+ *   &#64;Order(0)
+ *   static final S3ClassExtension S3 = S3ClassExtension.builder().createBucket("testbucket").build();
+ *
+ * }
  * </pre>
  */
-public class S3ClassExtension extends S3Mock implements BeforeAllCallback, AfterAllCallback {
+public class S3ClassExtension extends LocalS3Extension {
+
+  private final Set<String> buckets;
+  private final List<MockObject> mockObjects;
+  private LocalS3Endpoint classS3Endpoint;
+
+  private final S3Manager s3Manager = new S3Manager();
 
-  public S3ClassExtension(final List<String> buckets, final List<MockObject> mockObjects) {
-    super(buckets, mockObjects);
+  S3ClassExtension(Set<String> buckets, List<MockObject> mockObjects) {
+    this.buckets = buckets;
+    this.mockObjects = mockObjects;
   }
 
   @Override
-  public void beforeAll(ExtensionContext context) {
-    start();
+  public void beforeEach(ExtensionContext context) {
+    // this class should only be used as class level extension
+    // thus, not calling super
   }
 
   @Override
-  public void afterAll(ExtensionContext context) {
-    stop();
+  public void beforeAll(ExtensionContext context) throws Exception {
+    if (!context.getRequiredTestClass().isAnnotationPresent(LocalS3.class)) {
+      throw new IllegalStateException("The test class must be annotated with @LocalS3");
+    }
+
+    super.beforeAll(context);
+    Object portAsObject =
+        context
+            .getStore(ExtensionContext.Namespace.GLOBAL)
+            .get(context.getRequiredTestClass() + LOCAL_S3_PORT_STORE_SUFFIX);
+    if (portAsObject == null) {
+      throw new IllegalStateException("Could not determine port of S3 service");
+    }
+
+    var port = (int) portAsObject;
+    classS3Endpoint = new LocalS3Endpoint(port);
+    resetAll();
   }
 
-  @Override
-  public void start() {
-    super.start();
+  public void resetAll() {
+    try (S3Client client = newClient()) {
+      s3Manager.deleteAllBuckets(client);
+      s3Manager.createBuckets(client, buckets);
+      s3Manager.createMockObjects(client, mockObjects);
+    }
   }
 
-  @Override
-  public void stop() {
-    super.stop();
+  /**
+   * @return a new S3 client
+   * @deprecated Use {@link #newClient()} instead.
+   */
+  @Deprecated
+  public S3Client getClient() {
```

Comment:
Maybe it's just a matter of taste. It helps a little with the migration because it's one step less that users of the s3 testing module need to adapt right away. I've flagged it with `forRemoval` anyway. (I wonder who annotates code with `Deprecated` but does not want to remove it 🤔 )